back/engines/commercial/report_builder/app/services/report_builder/queries/device_types.rb

```diff
@@ -0,0 +1,44 @@
+module ReportBuilder
+  class Queries::DeviceTypes < ReportBuilder::Queries::Base
+    def run_query(
+      start_at: nil,
+      end_at: nil,
+      project_id: nil,
+      exclude_roles: nil,
+      **_other_props
+    )
+      start_date, end_date = TimeBoundariesParser.new(start_at, end_at).parse
+
+      sessions = ImpactTracking::Session.where(created_at: start_date..end_date)
+      sessions = apply_project_filter_if_needed(sessions, project_id)
+      sessions = exclude_roles_if_needed(sessions, exclude_roles)
+
+      counts_per_device_type = sessions
+        .where.not(device_type: nil)
+        .group(:device_type)
+        .count
+
+      {
+        counts_per_device_type: counts_per_device_type
+      }
+    end
+
+    def apply_project_filter_if_needed(sessions, project_id)
+      if project_id.present?
+        sessions_with_project = ImpactTracking::Pageview.where(project_id: project_id).select(:session_id)
```

Comment:
I think the select will trigger an extra query? It's tricky to fix, but I think something like this could work if the relations are in place in the model:
```
sessions.joins(:impact_tracking_pageviews).where(pageviews: { project_id: project_id })
```

Reply:
Good point, I was already thinking about something like this, but I was not sure if it would work since one session has many pageviews, and we want to keep the session where any of those pageviews has the project id. But maybe this will indeed work out of the box.

Since this logic is replicated across a few queries, I will try to fix it for all of them at the same time in the same ticket. Added it to the list here: https://www.notion.so/govocal/Other-various-visitor-data-tasks-1f29663b7b268040aaf8f92fc659fc05

Reply:
Yes indeed the many relationship complicates things, but normally a solution like this should work. A similar situation should be the `with_some_topics` scope in idea.rb:184 where an idea can have many `ideas_topics`.